compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -0,0 +1,914 @@
+
+import scala.annotation.tailrec
+
+import scala.collection.mutable
+
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.core._
+
+import Contexts._
+import Decorators._
+import Denotations._
+import Flags._
+import Names._
+import NameKinds.DefaultGetterName
+import Periods._
+import Phases._
+import StdNames._
+import Symbols._
+import SymDenotations._
+import Types._
+import TypeErasure.ErasedValueType
+
+import dotty.tools.dotc.transform.Erasure
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.Spans.Span
+import dotty.tools.dotc.report
+
+import org.scalajs.ir
+import org.scalajs.ir.{ClassKind, Position, Names => jsNames, Trees => js, Types => jstpe}
+import org.scalajs.ir.Names.{ClassName, MethodName, SimpleMethodName}
+import org.scalajs.ir.OriginalName
+import org.scalajs.ir.OriginalName.NoOriginalName
+import org.scalajs.ir.Position.NoPosition
+import org.scalajs.ir.Trees.OptimizerHints
+
+import dotty.tools.dotc.transform.sjs.JSSymUtils._
+
+import JSEncoding._
+
+final class JSExportsGen(jsCodeGen: JSCodeGen)(using Context) {
+  import jsCodeGen._
+  import positionConversions._
+
+  def genJSClassDispatchers(classSym: Symbol, dispatchMethodsNames: List[JSName]): List[js.MemberDef] = {
+    dispatchMethodsNames.map(genJSClassDispatcher(classSym, _))
+  }
+
+  private def genJSClassDispatcher(classSym: Symbol, name: JSName): js.MemberDef = {
+    val alts = classSym.info.membersBasedOnFlags(required = Method, excluded = Bridge)
+      .map(_.symbol)
+      .filter { sym =>
+        /* scala-js#3939: Object is not a "real" superclass of JS types.
+         * as such, its methods do not participate in overload resolution.
+         * An exception is toString, which is handled specially in genExportMethod.
+         */
+        sym.owner != defn.ObjectClass && sym.jsName == name
+      }
+      .toList
+
+    assert(!alts.isEmpty, s"Ended up with no alternatives for ${classSym.fullName}::$name.")
+
+    val (propSyms, methodSyms) = alts.partition(_.isJSProperty)
+    val isProp = propSyms.nonEmpty
+
+    if (isProp && methodSyms.nonEmpty) {
+      val firstAlt = alts.head
+      report.error(
+          i"Conflicting properties and methods for ${classSym.fullName}::$name.",
+          firstAlt.srcPos)
+      implicit val pos = firstAlt.span
+      js.JSPropertyDef(js.MemberFlags.empty, genExpr(name)(firstAlt.sourcePos), None, None)
+    } else {
+      genMemberExportOrDispatcher(name, isProp, alts, static = false)
+    }
+  }
+
+  private def genMemberExportOrDispatcher(jsName: JSName, isProp: Boolean,
+      alts: List[Symbol], static: Boolean): js.MemberDef = {
+    withNewLocalNameScope {
+      if (isProp)
+        genExportProperty(alts, jsName, static)
+      else
+        genExportMethod(alts.map(Exported), jsName, static)
+    }
+  }
+
+  def genJSConstructorDispatch(alts: List[Symbol]): (Option[List[js.ParamDef]], js.JSMethodDef) = {
+    val exporteds = alts.map(Exported)
```

Comment:
What is strange IMO is that we calculate things inside the class itself (rather than in the flow of export generation). An `Exported` has no more semantic meaning than a `Symbol` (it used to, but not anymore). So IMO a better thing would be something more semantically meaningful.

Reply:
So ... IIUC, it would be better to

1. Move the *computations* in a separate method that then instantiates the `Exported` with all the already computed data (rather than put everything in the constructor), and
2. Rename it to something more meaningful, perhaps `SymbolExportInfo`?

Would that address your concerns, or am I missing something?

Reply:
Yes. My feeling is that with a better control flow, some of this data becomes more local. But it's just a gut feeling. I haven't tried.